### PR TITLE
feat: Add toggle to show changes on save + confirm and reset changes

### LINKED ISSUE #407: feature request: Display config diff on save
### Checklist

- [x] I checked that this feature has not been requested before
- [x] I checked that this feature is not in the "Not planned" list
- [x] This feature will benefit the majority of users

### Problem Description / Use Case

Currently, when saving a configuration in aiostreams, changes are applied immediately without clear visibility into what has been modified. This can make it difficult to track adjustments, review unintended changes, or validate modifications before they are persisted. For users working with complex setups or sensitive parameters, this lack of transparency can lead to mistakes or misconfigurations that are hard to trace back.

### Proposed Solution

Introduce a confirmation step that displays a config diff before finalizing a save. When a user saves changes, aiostreams would:
- Open a window/dialog showing the difference between the previous config and the new one.
- Best case: Allow the user to view this diff either in a human-readable format (highlighting changed keys/values) or as JSON diff for exact inspection. Provide a toggle between views (human-readable vs raw JSON).
- Require user confirmation to apply the changes.

### PR DESCRIPTION
**Pull request by Crunging**
Added a toggle to show changes on save. When toggle is on it will show changes in a popup where you can confirm, cancel, and reset changes. This will greatly reduce people accidentally changing their configs and then thinking something is wrong. I also added some logic so it would play nice with my previous PR #646. I also added a check to show a prompt if no changes have been detected (will only show when toggle is on), and the config will still save just in case. Posted some examples of reordering addons, using my previous PR #646 to reorder group addon fetching, and what the new toggle and prompt looks like.

Closes #407 

Confirm_Changes
Confirm_Changes_2
No_Changes
Show_Changes

## Summary by CodeRabbit

* **New Features**
  * Preview configuration changes before saving with a two‑tab diff modal, navigable line diffs, Reset (revert), Confirm/Cancel, skip‑diff option and pre‑save diff check; readable value formatting and deterministic ordering.
* **Bug Fixes**
  * Removing a preset now also cleans up orphaned references across groups and addon lists.
* **Chores**
  * Optional user preference to remember the "Show changes" toggle and exposed default configuration for reuse.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: packages/frontend/src/components/menu/save-install.tsx [hunk 1]
```diff
@@ -145,23 +151,65 @@ function Content() {
...
+            return;
+          }
+        }
+      } catch (err) {
+        console.error('Error checking for changes:', err);
+        if (userData?.showChanges) {
+            toast.warning('Error checking for changes. Proceeding with save.');
+        }
+      }
+    }
```
[the remark this answers is not in the document]
Comment (Crunging) [reply packages/frontend/src/components/menu/save-install.tsx]: Addressed in commit c71a168

File: packages/frontend/src/utils/diff.ts [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+}
+
+export function getObjectDiff(
+  obj1: any,
+  obj2: any,
+  path: string[] = []
+): DiffItem[] {
+  const diffs: DiffItem[] = [];
+
+  const ignoredKeys = new Set(['uuid', 'trusted']);
```
[the remark this answers is not in the document]
Comment (Crunging) [reply packages/frontend/src/utils/diff.ts]: I have gone back and forth on this one. I have decided not to include it in the ignoredKeys as it is technically a user made change that is being saved to their config

[the remark this answers is not in the document]
Comment (Viren070) [reply packages/frontend/src/utils/diff.ts]: True, though that configuration option itself isn't really part of the aiostreams config and doesnt affect anything on stremio at all, its purely a setting in the configuration UI. 

It being stored in the user data should be abstracted to the user as its just there for convenience, if that makes sense. 

So I think adding it here makes sense.

File: packages/frontend/src/components/menu/save-install.tsx [hunk 3]
```diff
@@ -145,23 +151,67 @@ function Content() {
...
+      setLoading(true);
+      try {
+        const remoteResult = await UserConfigAPI.loadConfig(uuid, password);
+        
+        if (remoteResult.success && remoteResult.data) {
+          const diffs = getObjectDiff(remoteResult.data.config, userData);
+          
+          if (diffs.length === 0) {
+            toast.info('No changes detected');
+            suppressSuccessToast = true;
```
[the remark this answers is not in the document]
Comment (Crunging) [reply packages/frontend/src/components/menu/save-install.tsx:254]: This is intentional as when no changes are detected, it will go straight to the save operation where it is set to true right away.

File: packages/frontend/src/utils/diff.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Crunging) [reply packages/frontend/src/utils/diff.ts]: Sounds good, added it in

**Comment by Crunging** [comment]
Still working on some stuff, trying to see if this one quirk with groups can be fixed or not

**Comment by Crunging** [comment]
Okay I had to do a small rewrite on some portions to fix the quirk, but I think it creates a better result anyway. See pictures


Screenshot 2026-01-24 at 6 26 35 PM
Screenshot 2026-01-24 at 6 27 12 PM

**Comment by Crunging** [comment]
There's still a quirk present with the groups, but its a lot better than it was before and more manageable. For example, this is removing an addon from a group. It should be displayed as a change, but it is shown as an add and remove. This is the only quirk left that I have found. To my knowledge, to fix it would involve changing how the database stores groups or resigning to fuzzy logic. Everything else works as intended. 
Screenshot 2026-01-24 at 7 07 30 PM

File: packages/frontend/src/components/shared/diff-viewer.tsx [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+  if (diffs.length === 0) {
+    return (
+      <div className="text-center p-4 text-[--muted]">No changes detected.</div>
+    );
+  }
+
+  return (
+    <div className="w-full mt-4 max-h-[60vh] overflow-y-auto pr-2 custom-scrollbar">
+      <div className="space-y-3">
+        {diffs.map((diff, idx) => (
```
[the remark this answers is not in the document]
Comment (Crunging) [reply packages/frontend/src/components/shared/diff-viewer.tsx]: Implementing the fix below makes use of this variable

File: packages/frontend/src/components/menu/save-install.tsx [hunk 5]
```diff
@@ -369,6 +448,47 @@ function Content() {
...
+      const resolved = resolveDeep(val);
+
+      if (typeof resolved === 'object' && resolved !== null) {
+          try {
+              return JSON.stringify(resolved, null, 2);
+          } catch {
+              return '[Circular Reference]';
+          }
+      }
+      return String(resolved);
```
[the remark this answers is not in the document]
Comment (Crunging) [reply packages/frontend/src/components/menu/save-install.tsx]: This is on purpose as they are needed globally (addons, proxiedAddons, excludeAddons, etc.)

File: packages/frontend/src/components/menu/save-install.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (Crunging) [reply packages/frontend/src/components/menu/save-install.tsx:254]: I ended up changing this, I added setLoading(false)

File: packages/frontend/src/components/menu/addons.tsx
Comment (Crunging) [inline packages/frontend/src/components/menu/addons.tsx:1]: This needed to be added to prevent a bug where if you removed an addon from my addons, and the system auto removed it from the groups, it would still show as a ghost addon and a group change after multiple saves. This makes sure the groups stay consistent with the installed addon list.

Comment (Viren070) [reply packages/frontend/src/components/menu/addons.tsx:1]: This is because the invalid preset references get removed at the server during config validation:

https://github.com/Viren070/AIOStreams/blob/5f4737d61c32aeac9276b291604eba78cf33dfee/packages/core/src/utils/config.ts#L492-L525

We could probably copy this function over to the frontend and use it somewhere appropriate to handle the other cases too (title matching addons, year matching addons etc, ) 

The [applyMigrations function in the frontend](https://github.com/Viren070/AIOStreams/blob/5f4737d61c32aeac9276b291604eba78cf33dfee/packages/frontend/src/context/userData.tsx#L12-L165) was copied over rather than imported from [here](https://github.com/Viren070/AIOStreams/blob/5f4737d61c32aeac9276b291604eba78cf33dfee/packages/core/src/utils/config.ts#L527-L685) because otherwise the build failed for the frontend last time I tried, so you may need to do the same here.

File: packages/frontend/src/context/userData.tsx [hunk 6]
```diff
@@ -163,6 +163,41 @@ export function applyMigrations(config: any): UserData {
...
+  }
+  if (config.groups?.groupings) {
+    config.groups.groupings = config.groups.groupings.map((group) => ({
+      ...group,
+      addons: group.addons?.filter((addon) =>
+        existingPresetIds?.includes(addon)
+      ),
+    }));
+  }
+  return config;
```
[the remark this answers is not in the document]
Comment (Crunging) [reply packages/frontend/src/context/userData.tsx:199]: Not implemented to maintain parity with the server-side implementation. Refactoring to immutable might also introduce bugs

File: packages/frontend/src/components/menu/addons.tsx [hunk 7]
```diff
@@ -420,12 +421,14 @@ function Content() {
                                 setModalOpen(true);
                               }}
                               onRemove={() => {
-                                setUserData((prev) => ({
-                                  ...prev,
-                                  presets: prev.presets.filter(
+                                setUserData((prev) => {
+                                  if (!prev) return prev;
+                                  const cloned = structuredClone(prev);
```
[the remark this answers is not in the document]
Comment (Crunging) [reply packages/frontend/src/components/menu/addons.tsx:426]: Not implemented as we need a deep copy here as a shallow copy would cause the original data to be accidentally changed

File: packages/frontend/src/components/menu/addons.tsx [hunk 8]
```diff
@@ -420,12 +421,14 @@ function Content() {
                                 setModalOpen(true);
                               }}
                               onRemove={() => {
-                                setUserData((prev) => ({
-                                  ...prev,
-                                  presets: prev.presets.filter(
+                                setUserData((prev) => {
+                                  if (!prev) return prev;
```
Comment (Viren070) [inline packages/frontend/src/components/menu/addons.tsx:425]: why do we need to clone it?

Comment (Crunging) [reply packages/frontend/src/components/menu/addons.tsx:425]: From my understanding, we want to avoid mutating things with react, so we want to create a new object that replaces the old one with the changes.

https://react.dev/reference/eslint-plugin-react-hooks/lints/immutability

Comment (Crunging) [reply packages/frontend/src/components/menu/addons.tsx:425]: I am also using structuredClone because removeInvalidPresetReferences modifies nested properties. A shallow copy would not properly update these. I did it this way as I did not want to edit how removeInvalidPresetReferences works to keep it consistent with the backend

Comment (Viren070) [reply packages/frontend/src/components/menu/addons.tsx:425]: Ah okay, makes sense.

**Comment by Crunging** [comment]
Okay, this feature ended up being a lot more complicated than I initially thought. Every time after I fixed something and went to go test it, I would discover some new weird edge case. While each of these edge cases would have a simple workaround, I also knew if it was not addressed the support channel would be filled with questions. I do apologize if all these changes flooded your email. However, I think everything should be in a really good spot now, I also added a raw config viewer that will show changes if there ever happens to be a quirk with the simple UI change log.

**Comment by Viren070** [comment]
Thanks for the time and effort put into this PR, its a feature I have wanted in AIOStreams for a while. Flooding my emails is not a worry, no need to apologise. 

I will test this as soon as I can or just merge to main and let users test on nightly if I don't find time by the end of this week. 😛

**Comment by Crunging** [comment]
I realized that I forgot to key catalogs because I use AIOMetadata 😭

I just added a couple of lines for it, nothing changed. I promise this is still ready for review 😭
